### PR TITLE
Continue [Fix #1131 group retry degradation.]

### LINKED ISSUE #1131: Operations within "for_each>:" operator executed twice when retried
DIgdag version: 0.9.37

We have `_retry` parameter and `for_each` operator in our workflow.
When workflow failed, workflow itself successfully restarted.
But each operation in `_do` executed twice, even if the error came from outside of `for_each` operator.

this .dig file will reproduce the issue.

```
timezone: Asia/Tokyo

_retry: 2

+test:
  _export:
    counts: [
      0, 1, 2
    ]

  +echo:
    for_each>:
      count: ${counts}
    _do:
      echo>: ${count}  

  +failure:
    for_each>:
      count: ${counts}
    _do:
      sh>: exit ${count}
```

```
2019-06-04 14:41:25 +0900: Digdag v0.9.37
2019-06-04 14:41:36 +0900 [WARN] (main): Using a new session time 2019-06-04T00:00:00+09:00.
2019-06-04 14:41:36 +0900 [INFO] (main): Starting a new session project id=1 workflow name=z session_time=2019-06-04T00:00:00+09:00
2019-06-04 14:41:37 +0900 [INFO] (0016@[0:default]+z+test+echo): for_each>: {count=[0,1,2]}
2019-06-04 14:41:37 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=0=0): echo>: 0
0
2019-06-04 14:41:37 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=1=1): echo>: 1
1
2019-06-04 14:41:37 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=2=2): echo>: 2
2
2019-06-04 14:41:37 +0900 [INFO] (0016@[0:default]+z+test+failure): for_each>: {count=[0,1,2]}
2019-06-04 14:41:38 +0900 [INFO] (0016@[0:default]+z+test+failure^sub+for-0=count=0=0): sh>: exit 0
2019-06-04 14:41:38 +0900 [INFO] (0016@[0:default]+z+test+failure^sub+for-0=count=1=1): sh>: exit 1
2019-06-04 14:41:38 +0900 [ERROR] (0016@[0:default]+z+test+failure^sub+for-0=count=1=1): Task failed with unexpected error: Command failed with code 1
java.lang.RuntimeException: Command failed with code 1
        at io.digdag.standards.operator.ShOperatorFactory$ShOperator.runTask(ShOperatorFactory.java:143)
        at io.digdag.util.BaseOperator.run(BaseOperator.java:35)
        at io.digdag.core.agent.OperatorManager.callExecutor(OperatorManager.java:315)
        at io.digdag.cli.Run$OperatorManagerWithSkip.callExecutor(Run.java:705)
        at io.digdag.core.agent.OperatorManager.runWithWorkspace(OperatorManager.java:257)
        at io.digdag.core.agent.OperatorManager.lambda$runWithHeartbeat$2(OperatorManager.java:137)
        at io.digdag.core.agent.LocalWorkspaceManager.withExtractedArchive(LocalWorkspaceManager.java:25)
        at io.digdag.core.agent.OperatorManager.runWithHeartbeat(OperatorManager.java:135)
        at io.digdag.core.agent.OperatorManager.run(OperatorManager.java:119)
        at io.digdag.cli.Run$OperatorManagerWithSkip.run(Run.java:687)
        at io.digdag.core.agent.MultiThreadAgent.lambda$null$0(MultiThreadAgent.java:127)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2019-06-04 14:41:38 +0900 [INFO] (0016@[0:default]+z+test+echo): for_each>: {count=[0,1,2]}
2019-06-04 14:41:38 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=0=0): echo>: 0
0
2019-06-04 14:41:38 +0900 [INFO] (0018@[0:default]+z+test+echo^sub+for-0=count=0=0): echo>: 0
0
2019-06-04 14:41:39 +0900 [INFO] (0018@[0:default]+z+test+echo^sub+for-0=count=1=1): echo>: 1
1
2019-06-04 14:41:39 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=1=1): echo>: 1
1
2019-06-04 14:41:39 +0900 [INFO] (0016@[0:default]+z+test+echo^sub+for-0=count=2=2): echo>: 2
2
2019-06-04 14:41:39 +0900 [INFO] (0018@[0:default]+z+test+echo^sub+for-0=count=2=2): echo>: 2
2
2019-06-04 14:41:39 +0900 [INFO] (0018@[0:default]+z+test+failure): for_each>: {count=[0,1,2]}
2019-06-04 14:41:39 +0900 [INFO] (0016@[0:default]+z+test+failure^sub+for-0=count=0=0): sh>: exit 0
2019-06-04 14:41:39 +0900 [INFO] (0018@[0:default]+z+test+failure^sub+for-0=count=0=0): sh>: exit 0
2019-06-04 14:41:39 +0900 [INFO] (0018@[0:default]+z+test+failure^sub+for-0=count=1=1): sh>: exit 1
2019-06-04 14:41:39 +0900 [INFO] (0016@[0:default]+z+test+failure^sub+for-0=count=1=1): sh>: exit 1
2019-06-04 14:41:39 +0900 [ERROR] (0016@[0:default]+z+test+failure^sub+for-0=count=1=1): Task failed with unexpected error: Command failed with code 1
java.lang.RuntimeException: Command failed with code 1
        at io.digdag.standards.operator.ShOperatorFactory$ShOperator.runTask(ShOperatorFactory.java:143)
        at io.digdag.util.BaseOperator.run(BaseOperator.java:35)
        at io.digdag.core.agent.OperatorManager.callExecutor(OperatorManager.java:315)
        at io.digdag.cli.Run$OperatorManagerWithSkip.callExecutor(Run.java:705)
        at io.digdag.core.agent.OperatorManager.runWithWorkspace(OperatorManager.java:257)
        at io.digdag.core.agent.OperatorManager.lambda$runWithHeartbeat$2(OperatorManager.java:137)
        at io.digdag.core.agent.LocalWorkspaceManager.withExtractedArchive(LocalWorkspaceManager.java:25)
        at io.digdag.core.agent.OperatorManager.runWithHeartbeat(OperatorManager.java:135)
        at io.digdag.core.agent.OperatorManager.run(OperatorManager.java:119)
        at io.digdag.cli.Run$OperatorManagerWithSkip.run(Run.java:687)
        at io.digdag.core.agent.MultiThreadAgent.lambda$null$0(MultiThreadAgent.java:127)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

We observed this: 
- on client mode of version 0.9.35
- on client mode of version 0.9.37(used `digdag selfupdate`)
- on server mode of version 0.9.35

On the other hand, version 0.9.31 doesn't reproduce the issue; Each operation executed exactly once, when task retried (and this is what we expected, isn't it?).

### PR DESCRIPTION
**Pull request by naritta**
This PR will replace #1220 .
I added tests for only `if retry`. 
For checking backward-compatibility, there is already tests in https://github.com/treasure-data/digdag/blob/6ebb6230fb9be08f362e9eaad72349d9be79e0d5/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowExecutorTest.java#L165-L197 , so we should check if it succeeds or not.

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java [hunk 1]
```diff
@@ -905,16 +906,50 @@ public void addResumingTasks(long attemptId, List<ResumingTask> tasks)
...
+                }
+                String pfname = parentFullName.get();
+                String chfname = s.getFullName();
+                int pfnameLen = pfname.length();
+                int chfnameLen = chfname.length();
+
+                if (chfname.startsWith(pfname)
+                        && chfname.substring(pfnameLen).contains("^sub")) {
+                    // && chfnameLen > pfnameLen + "^sub".length()) {
+                    logger.debug("Exclude generated task:{} from retry target", chfname);
```
Comment (yoyama) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: These code may be better to remove.

**Comment by yoyama** [comment]
@naritta  I have one comment to code. Please confirm it.
I guess we should add more test cases. 
Especially this PR is try to fix #1131 and it reported on for_each> operator. 
I have also concern about call> operator and its cascade.

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java (hunk 1 above)
Comment (naritta) [reply digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: I deleted here. https://github.com/treasure-data/digdag/pull/1345/commits/1071cbf07425d0ddd1049138d0e696fcac9705c9

**Comment by naritta** [comment]
@yoyama Thank you for reviewing. I fixed and added some more tests, so could you check when you have time?

**Pull request merged**